### ISSUE TITLE
[BUG]: DAW UI Freeze when multiple instances are open

### ISSUE THREAD

**Issue by Franiac** [ZL-Audio/ZLEqualizer#142]
### Thanks for taking the time to fill out this bug report!

- [X] I have read the plugin manual and searched the existing issues.

### What happened?

```markdown
- Plugin version & format:

0.4.5 / VST 3

- A clear and concise description of what the bug is:

I found issue #137 and it is marked as 'not planned'. Well, I really enjoy ZL Equalizer a lot but if I open up more than 2 instances, the UI of the entire DAW (Reaper) gets laggy, 3 and 4 are like PowerPoint and the 5th instance completely kills Reaper and it is not coming back.

This seems like a serious issue in my eyes. Sure, there is the option to go 'Software Rendering' but I don't think that is the way to go in the otherwise functional OpenGL world of VST's.

The weird thing in my eyes is that neither the CPU or GPU spike to 100% so that it would make sense that the UI locks up. When I kill Reaper with ZL, the CPU is at 10% and the GPU at like 15%, so it's not an issue of overloading the GPU. Since nothing reacts anymore, my wild guess is that something blocks the event loop of the GUI thread.

I am a software dev myself. If you need debug info or anything, just tell me. If I can help, I gladly do. This is definitely a show-stopper for me and it could have a negative impact on new users (like me). I really want this plugin to blow up, it's just that good.

- A clear and concise description of what you expected to happen:

No matter the count of open ZL instances, the UI of the DAW should remain usable.
```

### Additional information

```markdown
- Operating system (& version if applicable):

Windows 11 Pro 24H2

- Hardware (CPU & GPU if applicable):

AMD Ryzen 9 5950X / Nvidia Quadro P4000

- DAW:

Reaper v7.28
```

### Code of Conduct

- [X] I agree to follow the Code of Conduct.

**Comment by zsliu98** [comment]
> I found issue #137 and it is marked as 'not planned'. Well, I really enjoy ZL Equalizer a lot but if I open up more than 2 instances, the UI of the entire DAW (Reaper) gets laggy, 3 and 4 are like PowerPoint and the 5th instance completely kills Reaper and it is not coming back.

As I said in the release note here https://github.com/ZL-Audio/ZLEqualizer/releases/tag/0.4.5, JUCE team is aware of this issue and they are working hard on a fix. Even some commercial plugins have to roll back to the software rendering because of the bad performance (on some machines). More details are available at this JUCE forum post: https://forum.juce.com/t/direct2d-causes-considerable-performance-drop/64009.

> Sure, there is the option to go 'Software Rendering' but I don't think that is the way to go in the otherwise functional OpenGL world of VST's.

Currently `software rendering` is the best solution if you are encountering the freezing issue.

> The weird thing in my eyes is that neither the CPU or GPU spike to 100% so that it would make sense that the UI locks up. When I kill Reaper with ZL, the CPU is at 10% and the GPU at like 15%, so it's not an issue of overloading the GPU. Since nothing reacts anymore, my wild guess is that something blocks the event loop of the GUI thread.

You are totally correct. JUCE 8 Direct2D painting takes too much time on the CPU, which finally blocks the GUI thread.

**Comment by Franiac** [comment]
I did not recognize the hint in the release notes. That said, I will wait for the JUCE team to fix the issue. After browsing through the forum post it seems like they really made a pretty big opsie. Can I leave this issue open to track whenever you are able to merge a fixed JUCE build?

**Comment by zsliu98** [comment]
Sure. I will leave this issue open until the fix is merged.

**Comment by zsliu98** [comment]
JUCE team has just released a new version. It seems that it won't block the GUI thread anymore. However, JUCE team is still working on Direct2D. See https://forum.juce.com/t/juce-roadmap-updates/62275/25. Hopefully, we will see further improvements in the future.

Would you like to try the following installer when you have time? Does it kill Reaper when you open multiple instances?

[ZL Equalizer-0.5.0-Windows.exe.zip](https://github.com/user-attachments/files/18352319/ZL.Equalizer-0.5.0-Windows.exe.zip)

**Comment by Franiac** [comment]
I tried your new build and here are my observations:

1. It is definitely better regarding the open instances. Before, 4-5 killed Reaper, now 10 are usabe. Everything above 10 is not usable anymore. 15-18 kill Reaper still. I mean, that's a lot more compared to 0.4.5 but given the fact I am able to open up 100+ IK Multimedia EQual instances without even noticing a performance drop, it's not that many.

2. I think ZL's refresh rate is lying 😄 (or I misinterpret it). I set it to 60Hz and compared it to EQual. Here is a little video recorded in 60fps to showcase this:

[https://franiac.com/zl.mkv](https://franiac.com/zl.mkv)

As you can see, EQual just looks way smoother and does not introduce that much lag behind the mouse pointer as ZL does. But here's the thing... if I set the refresh rate to 90 or even 120Hz, it changes nothing. Setting it to 25Hz definitely worsens the experience, but my feeling is that the refresh rate is capped at 30Hz. Or the refresh rate only applies to the spectrum but not the GUI controls itself? I don't know how the plugin renders its stuff so I might be wrong.

TLDR:

1. It's better, but still far behind "state of the art" plugins
2. GUI refresh rate seems to be capped at 30fps

Hope that helps 😄

**Comment by zsliu98** [comment]
> It is definitely better regarding the open instances. Before, 4-5 killed Reaper, now 10 are usabe. Everything above 10 is not usable anymore. 15-18 kill Reaper still.

I would assume that in 99% of cases, users will only open 2~3 EQ plugin windows and keep the remaining EQ plugin windows closed.

> I set it to 60Hz and compared it to EQual. Here is a little video recorded in 60fps to showcase this:

That refresh rate is the maximum refresh rate. The updating of the EQ curves is on another background thread. Therefore,

- if drawing one frame takes too long, the next frame will get dropped
- if the background thread cannot finish path calculation fast enough, the curves will not get updated immediately

Both will lead to an actual lower FPS.

I am still unsure how to update EQ curves on another thread without introducing some latency of updating on the GUI thread. It seems impossible to me. On macOS it is slightly better (but the latency is still there): https://drive.google.com/file/d/1xeUq0gYuoOkoQrJZsdWvC9PXoEmLkqvd/view?usp=sharing

Perhaps they are using OpenGL (does it help?). However, coding a complex GUI window in OpenGL is beyond my capability.

**Comment by Franiac** [comment]
Yes, 2-3 instances are the norm, I agree.

Ok, there's more the the refresh rate than I thought. I am glad I can trust my senses 😄 

Well, I think this issue is a general proof of the shortcomings of Direct2D vs. full OpenGL. You are absolutely right, IK uses OpenGL for their UI's, that's why these plugins are so smooth. The most manufacturers use ImGUI as their UI framework. It is immediate mode, extensively customizable and integrates wonderfully in an OpenGL render pipeline.

I appreciate the nice work on the ZL plugins and I will keep recommending them.
